google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsResumptionStrategy.java

```diff
@@ -0,0 +1,153 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.stub.readrows;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.retrying.StreamResumptionStrategy;
+import com.google.bigtable.v2.ReadRowsRequest;
+import com.google.bigtable.v2.ReadRowsRequest.Builder;
+import com.google.bigtable.v2.RowRange;
+import com.google.bigtable.v2.RowSet;
+import com.google.cloud.bigtable.data.v2.internal.ByteStringComparator;
+import com.google.cloud.bigtable.data.v2.models.RowAdapter;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+
+/**
+ * An implementation of a {@link StreamResumptionStrategy} for merged rows. This class tracks the
+ * last complete row seen and upon retry can build a request to resume the stream from were it left
+ * off.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications.
+ */
+@InternalApi
+public class ReadRowsResumptionStrategy<RowT>
+    implements StreamResumptionStrategy<ReadRowsRequest, RowT> {
+  private final RowAdapter<RowT> rowAdapter;
+  private ByteString lastKey = ByteString.EMPTY;
+  private long numProcessed;
+
+  public ReadRowsResumptionStrategy(RowAdapter<RowT> rowAdapter) {
+    this.rowAdapter = rowAdapter;
+  }
+
+  @Override
+  public boolean canResume() {
+    return true;
+  }
+
+  @Override
+  public StreamResumptionStrategy<ReadRowsRequest, RowT> createNew() {
+    return new ReadRowsResumptionStrategy<>(rowAdapter);
+  }
+
+  @Override
+  public void onProgress(RowT response) {
+    // Last key can come from both the last processed row key and a synthetic row marker. The
+    // synthetic row marker is emitted when the server has read a lot of data that was filtered out.
+    // So it can trim the start of the scan, but does not contribute to the row limit.
+    lastKey = rowAdapter.getKey(response);
+    if (!rowAdapter.isScanMarkerRow(response)) {
+      // Only real rows count towards the rows limit.
+      numProcessed++;
+    }
+  }
+
+  @Override
+  public ReadRowsRequest getResumeRequest(ReadRowsRequest request) {
+    if (lastKey.isEmpty()) {
+      return request;
+    }
+    Builder builder = request.toBuilder();
+
+    if (request.getRowsLimit() > 0) {
+      Preconditions.checkState(
+          request.getRowsLimit() >= numProcessed,
+          "Detected too many responses for the current row limit during a retry.");
+      builder.setRowsLimit(request.getRowsLimit() - numProcessed);
```

Comment:
If row limit is 10 and we have read 10 rows, do we need to continue from here? The new row limit will be 0, unlimited.

Reply:
If the limit is 10 and numProcessed is 10, then we are done. Could this happen? Make sure we have a test case for this.

Reply:
I updated it to `>`. I don't actually think this is practically possible. The server would have to send an error after sending all of the data. So I'm treating this as an illegal state & added a test for it

Reply:
Turns out I'm wrong and this could happen: https://github.com/GoogleCloudPlatform/google-cloud-go/commit/566ccf326f970b4fd1168bc0b6b474e94eeb6460#diff-37c5fcab0ce692b50ec93dc86e7a09d3.

The current code (after the change to`>`) is suboptimal because the stream will be terminated with an error despite having all of the rows, but it won't produce wrong results.  I don't see any easy way to fix this and think that this is a pretty rare edgecase that it might not be worth optimizing. I would like to pull this out as a separate issue to address separately.

Reply:
fixed